rust/ql/lib/codeql/rust/elements/internal/TypeInference.qll

```diff
@@ -0,0 +1,999 @@
+    TMethodTypeArgumentPosition(int pos) {
+      exists(any(MethodCallExpr mce).getGenericArgList().getTypeArgument(pos))
+    } or
+    TTypeParamTypeArgumentPosition(TypeParam tp)
+
+  class TypeArgumentPosition extends TTypeArgumentPosition {
+    int asMethodTypeArgumentPosition() { this = TMethodTypeArgumentPosition(result) }
+
+    TypeParam asTypeParam() { this = TTypeParamTypeArgumentPosition(result) }
+
+    string toString() {
+      result = this.asMethodTypeArgumentPosition().toString()
+      or
+      result = this.asTypeParam().toString()
+    }
+  }
+
+  class TypeParameterPosition = TypeParam;
+
+  bindingset[apos]
+  bindingset[ppos]
+  predicate typeArgumentParameterPositionMatch(TypeArgumentPosition apos, TypeParameterPosition ppos) {
+    apos.asTypeParam() = ppos
+    or
+    apos.asMethodTypeArgumentPosition() = ppos.getPosition()
+  }
+
+  private predicate id(Raw::TypeParam x, Raw::TypeParam y) { x = y }
+
+  private predicate idOfRaw(Raw::TypeParam x, int y) = equivalenceRelation(id/2)(x, y)
+
+  private int idOf(TypeParam node) { idOfRaw(Synth::convertAstNodeToRaw(node), result) }
+
+  int getTypeParameterId(TypeParameter tp) {
+    tp =
+      rank[result](TypeParameter tp0, int kind, int id |
+        tp0 instanceof RefTypeParameter and
+        kind = 0 and
+        id = 0
+        or
+        tp0 instanceof SelfTypeParameter and
+        kind = 0 and
+        id = 1
+        or
+        id = idOf(tp0.(TypeParamTypeParameter).getTypeParam()) and
+        kind = 1
+      |
+        tp0 order by kind, id
+      )
+  }
+}
+
+private import Input1
+
+private module M1 = Make1<Location, Input1>;
+
+private import M1
+
+class TypePath = M1::TypePath;
+
+module TypePath = M1::TypePath;
+
+private module Input2 implements InputSig2 {
+  private import TypeMention as TM
+
+  class TypeMention = TM::TypeMention;
+
+  TypeMention getABaseTypeMention(Type t) { result = t.getABaseTypeMention() }
+}
+
+private module M2 = Make2<Input2>;
+
+private import M2
+
+module Consistency = M2::Consistency;
+
+/** Gets the type annotation that applies to `n`, if any. */
+private TypeMention getTypeAnnotation(AstNode n) {
+  exists(LetStmt let |
+    n = let.getPat() and
+    result = let.getTypeRepr()
+  )
+  or
+  result = n.(SelfParam).getTypeRepr()
+  or
+  exists(Param p |
+    n = p.getPat() and
+    result = p.getTypeRepr()
+  )
+  or
+  exists(Function f |
+    result = f.getRetType().getTypeRepr() and
+    n = f.getBody()
+  )
+}
+
+/** Gets the type of `n`, which has an explicit type annotation. */
+pragma[nomagic]
+private Type resolveAnnotatedType(AstNode n, TypePath path) {
+  result = getTypeAnnotation(n).resolveTypeAt(path)
+}
+
+/**
+ * Holds if the type of `n1` at `path1` is the same as the type of `n2` at `path2`.
+ */
+bindingset[path1]
+bindingset[path2]
+private predicate typeSymmetry(AstNode n1, TypePath path1, AstNode n2, TypePath path2) {
```

Comment:
What about calling this `typeEquality`? That is more descriptive and matches the wording "the same" in the doc. Also this looks like the equality constraints in unification based type inference.